spec/support/expect_offense.rb

```diff
@@ -1,189 +1,163 @@
 # frozen_string_literal: true
 
+# Mixin for `expect_offense` and `expect_no_offenses`
+#
+# This mixin makes it easier to specify strict offense expectations
+# and a declarative and visual fashion. Just type out the code that
+# should generate a offense, annotate code by writing '^'s
+# underneath each character that should be highlighted, and follow
+# the carets with a string (separated by a space) that is the
+# message of the offense. You can include multiple offenses in
+# one code snippet.
+#
+# @example Usage
+#
+#     expect_offense(<<-RUBY.strip_indent)
+#       a do
+#         b
+#       end.c
+#       ^^^^^ Avoid chaining a method call on a do...end block.
+#     RUBY
+#
+# @example Equivalent assertion without `expect_offense`
+#
+#     inspect_source(cop, <<-RUBY.strip_indent)
+#       a do
+#         b
+#       end.c
+#     RUBY
+#
+#     expect(cop.offenses.size).to be(1)
+#
+#     offense = cop.offenses.first
+#     expect(offense.line).to be(3)
+#     expect(offense.column_range).to be(0...5)
+#     expect(offense.message).to eql(
+#       'Avoid chaining a method call on a do...end block.'
+#     )
+#
+# If you do not want to specify an offense then use the
+# companion method `expect_no_offenses`. This method is a much
+# simpler assertion since it just inspects the source and checks
+# that there were no offenses. The `expect_offenses` method has
+# to do more work by parsing out lines that contain carets.
 module ExpectOffense
   DEFAULT_FILENAME = 'example_spec.rb'.freeze
 
   # rubocop:disable Metrics/AbcSize
-  def expect_offense(source, filename: DEFAULT_FILENAME)
-    expectation = Expectation.new(source)
-    inspect_source(cop, expectation.source, filename)
-    offenses = cop.offenses.map(&method(:to_assertion)).sort
-
-    if expectation.assertions.empty?
-      raise 'Use expect_no_offenses to assert no violations'
+  def expect_offense(source, filename = DEFAULT_FILENAME)
+    expected_annotations = AnnotatedSource.parse(source)
+    if expected_annotations.plain_source == source
```

Comment:
I think i'd put a blank line on either side of this `if` and possibly another one before the final expectation as this method feels very claustrophobic.

Additionally, you might consider adding an `annotations?` (or similar) predicate to `AnnotatedSource` and push this logic down so that this method is slightly simpler. Just a suggestion though.